File: app/src/main/java/com/miandroidchallenge/ucoppp/miandroidchallenge/ui/deliverydetailsfragment/DeliveryDetailsFragment.java
package com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverydetailsfragment;

import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.miandroidchallenge.ucoppp.miandroidchallenge.R;
import com.miandroidchallenge.ucoppp.miandroidchallenge.databinding.FragmentDeliveryDetailsBinding;
import com.miandroidchallenge.ucoppp.miandroidchallenge.models.DeliveriesModel;
import com.miandroidchallenge.ucoppp.miandroidchallenge.ui.mainactivity.MainActivity;
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.ImageUtil;
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.NetworkUtil;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.Callable;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class DeliveryDetailsFragment extends Fragment implements OnMapReadyCallback {

    private static final String MARKER_IMAGE_NAME = "MARKER_IMAGE_NAME.png";
    private static final String KEY_DELIVERY = " delivery";
    private FragmentDeliveryDetailsBinding binding;

    private DeliveryDetailsViewModel viewmodel;

    private DeliveryDetailsFactoryViewModel factory;

    private GoogleMap googleMap;

    private DeliveriesModel deliveriesModel;

    private LatLng marker;

    public static DeliveryDetailsFragment newInstance(
            DeliveriesModel deliveriesModel
    ) {

        Bundle args = new Bundle();
        args.putParcelable(KEY_DELIVERY, deliveriesModel);
        DeliveryDetailsFragment fragment = new DeliveryDetailsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil
                .inflate(
                        inflater,
                        R.layout.fragment_delivery_details,
                        container,
                        false);

        deliveriesModel = getArguments().getParcelable(KEY_DELIVERY);

        binding.setDelivery(deliveriesModel);

        initArchitecture();

        initMap();

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (getActivity() instanceof MainActivity) {
            ((MainActivity) getActivity()).changeTitle(R.string.things_to_deliver);
        }
    }

    private void initArchitecture() {
        factory = new DeliveryDetailsFactoryViewModel(
                getActivity().getApplication()
        );

        viewmodel = ViewModelProviders
                .of(this, factory)
                .get(DeliveryDetailsViewModel.class);
    }

    private void initMap() {
        SupportMapFragment mMapFragment = (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map);

        if (mMapFragment != null) {
            mMapFragment.getMapAsync(this);
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;

        marker = new LatLng(
                deliveriesModel.getLocation().getLat(),
                deliveriesModel.getLocation().getLng()
        );

        setCameraPosition(marker);
        loadImage();
    }

    private void setCameraPosition(LatLng marker) {
        CameraPosition cameraPosition = CameraPosition
                .builder()
                .target(marker)
                .zoom(14f)
                .build();

        googleMap.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));
    }

    private void loadImage() {
        try {
            URL url = new URL(deliveriesModel.getImageUrl());

            String imagePath = ImageUtil.getImagePath(
                    getActivity().getApplication(),
                    MARKER_IMAGE_NAME);

            Observable.fromCallable((Callable<Bitmap>) () -> {
                if (!NetworkUtil.isConnected(getActivity().getApplication())) {
                    throw new Exception("Load local please");
                }
                return BitmapFactory.decodeStream(url.openConnection().getInputStream());
            })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Observer<Bitmap>() {
                        @Override
                        public void onSubscribe(Disposable d) {

                        }

                        @Override
                        public void onNext(Bitmap bitmap) {

                            ImageUtil.saveFile(
                                    getActivity().getApplication(),
                                    bitmap,
                                    MARKER_IMAGE_NAME
                            );

                            googleMap.addMarker(
                                    new MarkerOptions()
                                            .position(marker)
                                            .title(deliveriesModel.getDescription())
                                            .icon(BitmapDescriptorFactory.fromBitmap(bitmap)));

                        }

                        @Override
                        public void onError(Throwable e) {
                            googleMap.addMarker(
                                    new MarkerOptions()
                                            .position(marker)
                                            .title(deliveriesModel.getDescription())
                                            .icon(BitmapDescriptorFactory.fromBitmap(ImageUtil.getImage(imagePath))));
                        }

                        @Override
                        public void onComplete() {

                        }
                    });

        } catch (MalformedURLException e) {
            e.printStackTrace();
        }

    }
}

File: app/src/main/java/com/miandroidchallenge/ucoppp/miandroidchallenge/ui/deliverydetailsfragment/DeliveryDetailsFragment.kt
package com.miandroidchallenge.ucoppp.miandroidchallenge.ui.deliverydetailsfragment

import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.graphics.BitmapFactory
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.miandroidchallenge.ucoppp.miandroidchallenge.R
import com.miandroidchallenge.ucoppp.miandroidchallenge.databinding.FragmentDeliveryDetailsBinding
import com.miandroidchallenge.ucoppp.miandroidchallenge.models.Deliveries
import com.miandroidchallenge.ucoppp.miandroidchallenge.ui.mainactivity.MainActivity
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.getImage
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.getImagePath
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.isConnected
import com.miandroidchallenge.ucoppp.miandroidchallenge.util.saveFile
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.net.URL


class DeliveryDetailsFragment : Fragment(), OnMapReadyCallback {

    val MARKER_IMAGE_NAME = "MARKER_IMAGE_NAME.png"

    private lateinit var factory: DeliveryDetailsFactoryViewModel

    private lateinit var viewModel: DeliveryDetailsViewModel

    private lateinit var fragmentDeliveryDetailsBinding: FragmentDeliveryDetailsBinding

    private lateinit var googleMap: GoogleMap

    private lateinit var bundle: Deliveries

    private lateinit var marker: LatLng

    companion object {

        const val KEY_DELIVERY = "delivery"

        fun newInstance(delivery: Deliveries): DeliveryDetailsFragment {
            val fragment = DeliveryDetailsFragment()
            val bundle = Bundle()
            bundle.putParcelable(KEY_DELIVERY, delivery)
            fragment.arguments = bundle
            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        fragmentDeliveryDetailsBinding = DataBindingUtil
                .inflate(
                        inflater,
                        R.layout.fragment_delivery_details,
                        container,
                        false)

        bundle = arguments?.getParcelable(KEY_DELIVERY)!!

        fragmentDeliveryDetailsBinding.delivery = bundle

        initArchitecture()

        initMap()

        bundle = arguments?.getParcelable(KEY_DELIVERY)!!

        return fragmentDeliveryDetailsBinding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (activity is MainActivity) {
            (activity as MainActivity).changeTitle(R.string.things_to_deliver)
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        // Let's make this google map is global var
        this.googleMap = googleMap

        marker = LatLng(
                bundle.location?.lat!!,
                bundle.location?.lng!!
        )

        setCameraPosition(marker = marker)
        loadImage()

    }

    private fun initArchitecture() {
        factory = DeliveryDetailsFactoryViewModel(
                application = activity!!.application
        )

        viewModel = ViewModelProviders
                .of(this, factory)
                .get(DeliveryDetailsViewModel::class.java)
    }

    private fun initMap() {
        val mMapFragment =
                childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment

        mMapFragment.getMapAsync(this)
    }

    private fun loadImage() {
        val url = URL(bundle.imageUrl)

        val image = getImagePath(app = activity?.application!!,
                finalDirectory = MARKER_IMAGE_NAME
        )

        Observable
                .fromCallable({
                    if (!isConnected(activity?.application!!)) {
                        throw Exception("Load local please")
                    }
                    BitmapFactory.decodeStream(url.openConnection().getInputStream())
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ bitmap ->

                    saveFile(
                            app = activity?.application!!,
                            bitmap = bitmap,
                            imageName = MARKER_IMAGE_NAME
                    )

                    googleMap.addMarker(
                            MarkerOptions()
                                    .position(marker)
                                    .title(bundle.description)
                                    .icon(BitmapDescriptorFactory.fromBitmap(bitmap)))
                }, { _ ->
                    googleMap.addMarker(
                            MarkerOptions()
                                    .position(marker)
                                    .title(bundle.description)
                                    .icon(BitmapDescriptorFactory.fromBitmap(getImage(image))))
                })
    }

    private fun setCameraPosition(marker: LatLng) {
        val cameraPosition =
                CameraPosition
                        .builder()
                        .target(marker)
                        .zoom(14f)
                        .build()

        googleMap.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
    }
}